source/particle/property/viscoplastic_strain_invariants.cc

```diff
@@ -117,45 +117,24 @@ namespace aspect
           }
 
-        // Find out plastic yielding by calling function in material model.
-        const MaterialModel::ViscoPlastic<dim> &viscoplastic
-          = Plugins::get_plugin_as_type<const MaterialModel::ViscoPlastic<dim>>(this->get_material_model());
-
-        const bool plastic_yielding = viscoplastic.is_yielding(material_inputs);
-
-
-        /* Next take the integrated strain invariant from the prior time step. When
-         * there are two fields (plastic and viscous), this assumes plastic
-         * strain is always in data position one and viscous strain in position two.
-         * In this case old_strain will first be given the plastic strain, and then,
-         * if there is no plastic yielding it will update to the viscous strain instead.
-         */
-        auto &data = particle->get_properties();
-        double old_strain = data[data_position];
-        if (n_components == 2 && plastic_yielding == false)
-          old_strain = data[data_position+(n_components-1)];
-
-
-        // Calculate strain rate second invariant
-        const double edot_ii = std::sqrt(std::fabs(second_invariant(deviator(material_inputs.strain_rate[0]))));
+        // Evaluate directly in the viscoplastic material model and modify the reaction outputs
+        this->get_material_model().evaluate (material_inputs,material_outputs);
 
-        // New strain is the old strain plus dt*edot_ii
-        const double new_strain = old_strain + dt*edot_ii;
-
-
-        /* Once we know whether the particle underwent plastic, viscous, or
-         * total strain assign the new strain to the correct data position.
-         * NOTE: This assumes that total strain cannot be used in combination with the
-         * other fields. If this changes in the future, this will need to be updated.
-         * */
-        if (this->introspection().compositional_name_exists("plastic_strain") && plastic_yielding == true)
-          data[data_position] = new_strain;
-
-        if (this->introspection().compositional_name_exists("viscous_strain") && plastic_yielding == false)
-          data[data_position+(n_components-1)] = new_strain;
+        const int plastic_strain_index = this->introspection().compositional_index_for_name("plastic_strain"); 
+        const int viscous_strain_index = this->introspection().compositional_index_for_name("viscous_strain");
+        const int total_strain_index = this->introspection().compositional_index_for_name("total_strain");
+        
+        if (this->introspection().compositional_name_exists("plastic_strain"))
+          particle->get_properties()[data_position] += material_outputs.reaction_terms[0][plastic_strain_index];
+        
+        if (this->introspection().compositional_name_exists("viscous_strain"))
+          particle->get_properties()[data_position] += material_outputs.reaction_terms[0][viscous_strain_index];
```

Comment:
This can now be the following:

```suggestion
        unsigned int particle_property = 0;
        if (plastic_strain_index != numbers::invalid_unsigned_int)
        {
          particle->get_properties()[data_position+i] += material_outputs.reaction_terms[0][plastic_strain_index];
          ++i;
        }
        
        if (viscous_strain_index != numbers::invalid_unsigned_int)
        {
          particle->get_properties()[data_position+i] += material_outputs.reaction_terms[0][viscous_strain_index];
          ++i;
         }
          
        if (total_strain_index != numbers::invalid_unsigned_int)
        {
          particle->get_properties()[data_position+i] += material_outputs.reaction_terms[0][total_strain_index];
          ++i;
          }
```